Proposal: Granular Endpoint Configuration (#5505)
thanos-io/thanos: docs/proposals-accepted/202207-granular-endpoint-configuration.md

---
type: proposal
title: Granular Endpoint Configuration
status: Approved
owner: SrushtiSapkale
menu: proposals-accepted
---

* **Related Tickets:**

  * [Per-Store TLS configuration in Thanos Query](https://github.com/thanos-io/thanos/issues/977)

## What

> This proposal builds up on the [Unified Endpoint Discovery Proposal](https://thanos.io/tip/proposals-accepted/202101-endpoint-discovery.md/) and [Automated, per-endpoint mTLS Proposal](https://thanos.io/tip/proposals-accepted/202106-automated-per-endpoint-mtls.md/)

Reviewer: ```suggestion
This proposal builds up on the [Unified Endpoint Discovery Proposal](https://thanos.io/tip/proposals-accepted/202101-endpoint-discovery.md/) and [Automated, per-endpoint mTLS Proposal](https://thanos.io/tip/proposals-accepted/202106-automated-per-endpoint-mtls.md/) to add possibility to have granular setting per endpoint and deprecate old flags.
```

## Why

> The Thanos Querier component supports basic mTLS configuration for internal gRPC communication. Mutual TLS (mTLS) ensures that traffic is both secure and trusted in both directions between a client and server. This works great for basic use cases but it still requires extra forward proxies to make it work for bigger deployments.

Reviewer: Let's motivate the goals here. (:

> The Thanos Querier component supports basic mTLS configuration for internal gRPC communication. Mutual TLS (mTLS) ensures that traffic is both secure and trusted in both directions between a client and server. This works great for basic use cases but it still requires extra forward proxies to make it work for bigger deployments.

Reviewer: ```suggestion
The Thanos Querier component supports basic mTLS configuration for internal gRPC communication. Mutual TLS (mTLS) ensures that traffic is both secure and trusted in both directions between a client and server. This works great for basic use cases, but it still requires extra forward proxies to make it work for quite common deployments where different servers have different auth, cert or TLS options.

This is similar to cases where we want to have further per-endpoint configuration pieces like what APIs to enable or in future what load balancing groups it belongs to.

We also have to be prepared for more options that will appear as the functionality of Queriers grows per endpoint (e.g limits).
```

> Let’s imagine we have an Observer Cluster that is hosting Thanos Querier along with Thanos Store Gateway. Within the same Observer Cluster, we would like to connect one or more Thanos Sidecars. Additionally, we also want to connect the Querier in the Observer Cluster to several remote instances of Thanos Sidecar in remote clusters.
> 
> In this scenario we can use a proxy server (e.g. envoy). In thanos as we have multiple queriers connected to a central querier, Envoy can efficiently loadbalance gRPC connections, something we currently don't support natively in Thanos. If we used the endpoints config to connect the global to the regional queriers, they will not be able to scale regional queriers due to the persistent nature of gRPC connections. We can use Envoy when we need to scale our regional queries but here having an endpoint configuration is a better option in distributed store architecture as Thanos can communicate securely and directly through remote endpoints.

Reviewer: ```suggestion

```

## Pitfalls of the current solution

> Ideally, we would want to use mTLS to encrypt the connection to the remote clusters. If we would enable the current mTLS, it would be applied to all the storeAPI’s but we don’t want it to be applied on the storeAPI’s of central Thanos instance (Observer cluster) in which Thanos query component is present (for faster communications with storeAPI’s (sidecars) of same cluster, to reduce the pain of provisioning certificates etc.). So it requires extra forward proxies to make it work for bigger (multi-cluster) deployments.

Reviewer: ```suggestion
Current users cannot use different certifcates, auth or TLS options for different endpoints. The current workarounds is to use layered queries for this which is much more complex and confusing deployment.

Second, quite viable option is to use forward proxy like envoy. The problem with this approach is that any Thanos adopter have to learn yet another configuration language/options, deployment (build, release, support) models and operational (metrics, migration) models  which is not trivial and hinders adoptions.
```

## Goals

> * To add support for per-endpoint TLS configuration in Thanos Query Component for internal gRPC communication.

Reviewer: ```suggestion
* To add support for per-endpoint granularity configuration. For example:
   * TLS configuration in for Querier gRPC communication.
   * Basic/Token Auth
   * Strict mode
   * Enabled APIs (specifying that server which can serve Exemplars and Store should only be asked for Exemplars)
   * (perhaps more in future)

   
```

> * Adding a new `endpoint.config` flag will deprecate the following flags: `store.sd-interval`, `store.sd-dns-interval`, `store.sd-dns-resolver`, `store.sd-files` and all `grpc-client-.*`. We can mark the current flags as deprecated, and after some time we can remove them.

Reviewer: ```suggestion
* Unify endpoint configuration. Which means deprecating all non `endpoint` flags for endpoint configuration like `store.sd-interval`, `store.sd-dns-interval`, `store.sd-dns-resolver`, `store.sd-files` and all `grpc-client-.*`.
```

> * Enable token based auth for endpoints

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion

## Non Goals

* Granular configuration & implementation for load balancing groups. This can be improved in the next proposals and is currently fulfilled using any forward or reverse proxy like `envoy`, `caddy` or `nginx`.
```

## How

A new CLI option `--endpoints.config`, with no dynamic reloading, which will accept the path to a yaml file is proposed which contains a list as follows :

> ```yaml

Reviewer: I wonder if we want to tacle the @yeya24 request to enable different API

Reviewer: I or @saswatamcode will add more context

> - tls_config:

Reviewer: I think we need some root object for easier to understand structure and possibility to add more "global" fields 
```suggestion
endpoints:
- tls_config:
```

    cert_file: ""
    key_file: ""
    ca_file: ""
    server_name: ""
  endpoints: []
  endpoints_sd_files:
    - files: []
  mode: ""
```

> The YAML file contains set of endpoints (e.g Store API) with optional TLS options. To enable TLS either use this option or deprecated ones --grpc-client-tls* .

Reviewer: Can I specify multiple `endpoint.config`  flags?

Reviewer: cc me @saswatamcode

If the mode is strict (i.e. `mode: ”strict”`) then all the endpoints in that list item will be considered as strict (statically configured store API servers that are always used, even if the health check fails, as supplied in `--endpoint.strict`). And if there is no mode specified (i.e `mode: “”`) then all endpoints in that item will be considered normal.

Also we noticed that even if the user wants to connect to the store api only, the store method indiscriminately returns all endpoints ie exemplar api, store api, etc, even if they do not expose store API which is confusing. This is because the query exposes endpoints of all the apis. We can update the filtering logic where the querier endpoint knows which api endpoints it should expose for a particular component.

## Alternatives

> * Granular endpoint configuration is not a solution for federate querier loadbalncing and scalability. As an alternative can embed Envoy in thanos for loadbalancing. This would work efficiently because endpointset cannot configure federated queriers when they scale due to persistent nature of grpc connections.

Reviewer: What about moving this to NonGoals

If the mode is strict (i.e. `mode: ”strict”`) then all the endpoints in that list item will be considered as strict (statically configured store API servers that are always used, even if the health check fails, as supplied in `--endpoint.strict`). And if there is no mode specified (i.e `mode: “”`) then all endpoints in that item will be considered normal.

## TODO

* To update the filtering of APIs in the querier to return only the endpoints that are needed.

For example:
The user wants to connect to the store api only.But here, we notice that the store method indiscriminately returns all endpoints ie exemplar api, store api, etc, even if they do not expose store API.

This is because the query exposes endpoints of all the apis. We can update the filtering logic where the querier endpoint knows which api endpoints it should expose for a particular component.

## Alternatives

## Action Plan

* Implement `--endpoint.config`.

> * Adding a new `endpoint.config` flag will deprecate the following flags: `store.sd-interval`, `store.sd-dns-interval`, `store.sd-dns-resolver`, `store.sd-files` and all `grpc-client-.*`. We can also remove the hidden flags `ruleEndpoints`, `metadataEndpoints`, `exemplarEndpoints`.

Reviewer 2: So I think the hidden flags are a bit different here, i.e, these are the flag variables not the flags themselves for e.g, `--rule` not `ruleEndpoints`.
